protocol/communities/community_event.go

```diff
@@ -0,0 +1,430 @@
+	return &CommunityEvent{
+		CommunityEventClock: o.NewCommunityEventClock(),
+		Applied:             true,
+		Type:                protobuf.CommunityEvent_COMMUNITY_MEMBER_TOKEN_PERMISSION_DELETE,
+		TokenPermission:     permission,
+		CommunityConfig: &protobuf.CommunityConfig{
+			Encrypted: o.Encrypted(),
+		},
+	}
+}
+
+func (o *Community) ToCommunityRequestToJoinAcceptCommunityEvent(changes *CommunityEventChanges) *CommunityEvent {
+	return &CommunityEvent{
+		CommunityEventClock:    o.NewCommunityEventClock(),
+		Applied:                true,
+		Type:                   protobuf.CommunityEvent_COMMUNITY_REQUEST_TO_JOIN_ACCEPT,
+		MembersAdded:           changes.MembersAdded,
+		AcceptedRequestsToJoin: changes.AcceptedRequestsToJoin,
+	}
+}
+
+func (o *Community) ToCommunityRequestToJoinRejectCommunityEvent(changes *CommunityEventChanges) *CommunityEvent {
+	return &CommunityEvent{
+		CommunityEventClock:    o.NewCommunityEventClock(),
+		Applied:                true,
+		Type:                   protobuf.CommunityEvent_COMMUNITY_REQUEST_TO_JOIN_REJECT,
+		RejectedRequestsToJoin: changes.RejectedRequestsToJoin,
+	}
+}
+
+func (o *Community) UpdateCommuntyByEvents(communityEventMessage *CommunityEventsMessage) (*CommunityChanges, error) {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	if communityEventMessage.CommunityDescriptionClock != o.config.CommunityDescription.Clock {
+		return nil, errors.New("clock for admin event message is outdated")
+	}
+
+	// Merge community admin events to existing community. Admin events must be stored to the db
+	// during saving the community
+	o.mergeCommunityEvents(communityEventMessage.Events)
+
+	// Create a deep copy of current community so we can update CommunityDescription by new admin events
+	copy := o.createDeepCopy()
+
+	// Update the copy of the Community with the new community events
+	err := copy.updateCommuntyDescriptionByNewEvents()
+	if err != nil {
+		return nil, err
+	}
+
+	// Collect `CommunityChanges` data by searching a difference between `CommunityDescrption`
+	// with applied events and `CommunityDescrption` with new events
+	changes, err := o.collectCommuntyChanges(copy.config.CommunityDescription)
+	if err != nil {
+		return nil, err
+	}
+
+	changes.Community = copy
+
+	return changes, nil
+}
+
+func (o *Community) updateCommuntyDescriptionByNewEvents() error {
+	prevAdminClock := o.config.CommunityDescription.Clock + 1
+	for i := range o.config.Events {
+		communityEvent := &o.config.Events[i]
+		if !communityEvent.Applied {
+			err := validateCommunityEvent(communityEvent)
+			if err != nil {
+				return err
+			}
+			err = o.updateCommuntyDescriptionByCommunityEvent(*communityEvent, prevAdminClock)
+			if err != nil {
+				return err
+			}
+		}
+		prevAdminClock = communityEvent.CommunityEventClock
+		communityEvent.Applied = true
+	}
+	return nil
+}
+
+func (o *Community) updateCommuntyDescriptionByAppliedEvents() error {
+	prevAdminClock := o.config.CommunityDescription.Clock + 1
```

Comment:
Is this being increased by 1 to ensure it's not equal to the previous one so it'll be processed properly?
If yes, or any other specific reason, I'd add a comment about that here so others will know when they run into this line.